sdk/eventhub/azure-eventhubs/azure/eventhub/aio/_producer_client_async.py

```diff
@@ -0,0 +1,179 @@
+        self._max_message_size_on_link = 0
+
+    async def _init_locks_for_producers(self):
+        if not self._producers:
+            async with self._client_lock:
+                if not self._producers:
+                    num_of_producers = len(await self.get_partition_ids()) + 1
+                    self._producers = [None] * num_of_producers
+                    for i in range(num_of_producers):
+                        self._producers_locks.append(asyncio.Lock())
+                        # self._producers_locks = [asyncio.Lock()] * num_of_producers
+
+    async def send(self, event_data: Union[EventData, EventDataBatch, Iterable[EventData]],
+            *, partition_key: Union[str, bytes] = None, partition_id: str = None, timeout: float = None) -> None:
+        """Sends an event data and blocks until acknowledgement is received or operation times out.
+
+        :param event_data: The event(s) to be sent. It can be an EventData object, an iterable of EventData objects
+            or an EventDataBatch, which contains some EventData objects. EventDataBatch is highly recommended to group
+            as many EventData objects as possible in a single send call. This will maximize the send throughput.
+            A send operation will send all EventData objects to one partition.
+            Multiple send operations will round-robin send to partitions alternately.
+            Round-robin is the recommended and the default behavior of send
+            unless you set partition_id or partition_key.
+        :param partition_id: event_data will be sent to this partition. This is not recommended because it may
+         cause unbalanced number of event data across partitions.
+        :param partition_key: With the given partition_key, event_data will land to
+         a particular partition but unknown until the send completes. Two send operations with the same partition_key
+         will go to the same partition even though which the partition is unknown while before send.
+        :param timeout: The maximum wait time to send the event data.
+         Default value is 60 seconds. If set to 0, there will be no timeout.
+
+        :raises: ~azure.eventhub.AuthenticationError, ~azure.eventhub.ConnectError, ~azure.eventhub.ConnectionLostError,
+                ~azure.eventhub.EventDataError, ~azure.eventhub.EventDataSendError, ~azure.eventhub.EventHubError
+        :return: None
+        :rtype: None
+
+        Example:
+            .. literalinclude:: ../samples/test_examples_eventhub_async.py
+                :start-after: [START eventhub_producer_client_send_async]
+                :end-before: [END eventhub_producer_client_send_async]
+                :language: python
+                :dedent: 4
+                :caption: Asynchronously sends an event data
+
+        """
+
+        await self._init_locks_for_producers()
+
+        producer_index = int(partition_id) if partition_id is not None else -1
```

Comment:
Is it contractual that `partition_id`s range from `0` to <number of partitions> - 1? And why do we treat partition ids as strings everywhere in the public api surface area if that is the case?

Reply:
According to service team, partition_id is always from 0 to N even though we (all languages) decided to use string. My understanding is to avoid potential break change some day to our users if service team decided to change it to string. Not likely to happen though.
"send" method is called many times so I chose to use list instead of dict for it's better performance.
I've no concern to change it to dict to be consistent.

Reply:
If the service team changes that assumption, this logic has to very much change, right? In other words, you are currently **not** using it a/assuming it is an arbitrary string.

Reply:
did yo actually that dict are lower performance than list in that scenario? there has been numerous improvment on dict performance over the year, and I won't be surprised this assumption is not true on recent Python

Reply:
Points are taken. Since we have started the long running test, I'm not going to change this part at this time to risk new bugs introduced at this time. This is internal implementation. No API impact to users.
I'll change it after preview5. I don't forsee any possibilities that service is changed before GA or any time soon from what I understand.

Reply:
Not blocking for preview5, solving